testing/src/test/java/io/stargate/it/storage/StargateContainer.java

```diff
@@ -467,33 +497,26 @@ public String rack() {
   }
 
   private static class Env {
-
-    private final List<Integer> ports = new ArrayList<>();
-    private final int addressStart;
-
-    private Env(int nodeCount) throws IOException {
-      // Note: do not reuse addresses
-      addressStart = stargateAddressStart.getAndAdd(nodeCount);
-
-      // Allocate `nodeCount` random ports
-      List<ServerSocket> sockets = new ArrayList<>();
-      for (int i = 0; i < nodeCount; i++) {
-        ServerSocket socket = new ServerSocket(0);
-        sockets.add(socket);
-        ports.add(socket.getLocalPort());
-      }
-
-      for (ServerSocket socket : sockets) {
-        socket.close();
-      }
-    }
+    private final Map<Integer, Integer> jmxPorts = new ConcurrentHashMap<>();
+    private final Map<Integer, String> listenAddresses = new ConcurrentHashMap<>();
 
     private String listenAddress(int index) {
-      return "127.0.0." + (addressStart + index);
+      return listenAddresses.computeIfAbsent(
+          index, i -> "127.0.0." + stargateAddressStart.getAndIncrement());
     }
 
     private int jmxPort(int index) {
-      return ports.get(index);
+      return jmxPorts.computeIfAbsent(
+          index,
+          i -> {
+            try {
+              try (ServerSocket socket = new ServerSocket(0)) {
+                return socket.getLocalPort();
+              }
+            } catch (IOException e) {
+              throw new UncheckedIOException(e);
+            }
+          });
```

Comment:
I think this is a bit risky. We immediately release the socket, so there's no strict guarantee that the next call will not chose the same port, I guess :) This may be critical when we start multi-node Stargate clusters all at once.

Reply:
I also think preallocating is safer, we can define an upper bound (like 10 nodes) for all tests and work with that.

Reply:
Good catch. I'll pre-allocate.